beacon_node/beacon_chain/src/validator_monitor.rs

```diff
@@ -489,6 +532,130 @@ impl<T: EthSpec> ValidatorMonitor<T> {
                 }
             }
         }
+
+        // Add missed non-finalized blocks for the monitored validators
+        // count the amount of times a monitored validator missed a non-finalized block
+        self.missed_blocks
+            .iter()
+            .group_by(|(_, validator_index, _)| *validator_index)
+            .into_iter()
+            .for_each(|(validator_index, group)| {
+                let missed_blocks_count = group.count();
+                self.aggregatable_metric(validator_index.to_string().as_str(), | label| {
+                    metrics::set_int_gauge(
+                        &metrics::VALIDATOR_MONITOR_MISSED_NON_FINALIZED_BLOCKS_AT_NON_FINALIZED_EPOCH_TOTAL,
+                        &[label],
+                        u64_to_i64(missed_blocks_count as u64),
+                    );
+                });
+            });
+
+        // add to the total missed blocks by monitored validator
+        self.total_missed_blocks
+            .iter()
+            .for_each(|(i, c)| {
+                self.aggregatable_metric(i.to_string().as_str(), |label| {
+                    metrics::set_int_gauge(
+                        &metrics::VALIDATOR_MONITOR_MISSED_NON_FINALIZED_BLOCKS_TOTAL,
+                        &[label],
+                        *c as i64,
+                    );
+                });
+            });
+    }
+
+    /// Add missed non-finalized blocks for the monitored validators
+    fn add_validators_missed_blocks(&mut self, state: &BeaconState<T>) {
+        // Define range variables
+        let current_slot = state.slot();
+        let range_of_slots = T::slots_per_epoch() as usize - MISSED_BLOCK_LAG_SLOTS;
+        let start_slot = current_slot - Slot::new(range_of_slots as u64);
+
+        // As we are skipping the genesis slot, we can simply pass Hash256::zero() as the shuffling decision block
+        // cf. state.proposer_shuffling_decision_root_at_epoch implementation
+        let shuffling_decision_block = Hash256::zero();
+
+        // List of proposers per epoch from the beacon_proposer_cache
+        let mut proposers_per_epoch: Option<SmallVec<[usize; TYPICAL_SLOTS_PER_EPOCH]>> = None;
+        for n in 1..range_of_slots {
```

Comment:
I'm unsure of the math in this loop. I made a sketch of it in this [Rust Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=12f71d5ef23897318364161e0bc23d15) and it's only visiting 23 slots when I'd expect it to visit 28 (`slots_per_epoch - missed_block_lag_slots`).

I've run out of time to suggest an alterative, would you mind having a look please?

Reply:
I changed the loop as one of the edge cases would make me do some if/else to dertemine whether we should get inside the loop or not. Should b more elegant that way. Now, it's going top to bottom. I used your playground [here](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=6c58503152a5402958ea1d136d8e56e5) with different scenarios.